File: hexameter.core/src/main/java/org/codetome/hexameter/core/internal/impl/HexagonalGridImpl.java
package org.codetome.hexameter.core.internal.impl;

import org.codetome.hexameter.core.api.*;
import org.codetome.hexameter.core.api.contract.HexagonDataStorage;
import org.codetome.hexameter.core.api.contract.SatelliteData;
import org.codetome.hexameter.core.backport.Optional;
import org.codetome.hexameter.core.internal.GridData;

import java.util.*;

public final class HexagonalGridImpl<T extends SatelliteData> implements HexagonalGrid<T> {

    private static final int[][] NEIGHBORS = {{+1, 0}, {+1, -1}, {0, -1}, {-1, 0}, {-1, +1}, {0, +1}};
    private static final int NEIGHBOR_X_INDEX = 0;
    private static final int NEIGHBOR_Z_INDEX = 1;

    private final GridData gridData;
    private final HexagonDataStorage<T> hexagonDataStorage;

    /**
     * Creates a new HexagonalGrid based on the provided HexagonalGridBuilder.
     *
     * @param builder builder
     */
    public HexagonalGridImpl(final HexagonalGridBuilder<T> builder) {
        this.gridData = builder.getGridData();
        this.hexagonDataStorage = builder.getHexagonDataStorage();
        for (CubeCoordinate cubeCoordinate : builder.getGridLayoutStrategy().fetchGridCoordinates(builder)) {
            HexagonalGridImpl.this.hexagonDataStorage.addCoordinate(cubeCoordinate);
        }
    }

    @Override
    public GridData getGridData() {
        return gridData;
    }

    @Override
    public Iterable<Hexagon<T>> getHexagons() {

        final Iterator<CubeCoordinate> coordIter = hexagonDataStorage.getCoordinates().iterator();

        return new Iterable<Hexagon<T>>() {
            @Override
            public Iterator<Hexagon<T>> iterator() {
                return new Iterator<Hexagon<T>>() {
                    @Override
                    public boolean hasNext() {
                        return coordIter.hasNext();
                    }

                    @Override
                    public void remove() {
                        throw new UnsupportedOperationException("This is not a mutable iterator.");
                    }

                    @Override
                    public Hexagon<T> next() {
                        return new HexagonImpl<>(gridData, coordIter.next(), hexagonDataStorage);
                    }
                };
            }
        };
    }

    @Override
    public Iterable<Hexagon<T>> getHexagonsByCubeRange(final CubeCoordinate from, final CubeCoordinate to) {
        final List<CubeCoordinate> coordinates = new ArrayList<>();

        for (int gridZ = from.getGridZ(); gridZ <= to.getGridZ(); gridZ++) {
            for (int gridX = from.getGridX(); gridX <= to.getGridX(); gridX++) {
                coordinates.add(CubeCoordinate.fromCoordinates(gridX, gridZ));
            }
        }

        final Iterator<CubeCoordinate> iter = coordinates.iterator();
        while (iter.hasNext()) {
            CubeCoordinate next = iter.next();
            if (!containsCubeCoordinate(next)) {
                iter.remove();
            }
        }

        final Iterator<CubeCoordinate> coordIter = coordinates.iterator();

        return new Iterable<Hexagon<T>>() {
            @Override
            public Iterator<Hexagon<T>> iterator() {
                return new Iterator<Hexagon<T>>() {
                    @Override
                    public boolean hasNext() {
                        return coordIter.hasNext();
                    }

                    @Override
                    public Hexagon<T> next() {
                        return getByCubeCoordinate(coordIter.next()).get();
                    }

                    @Override
                    public void remove() {
                        throw new UnsupportedOperationException("This is not a mutable iterator.");
                    }
                };
            }
        };
    }

    @Override
    public Iterable<Hexagon<T>> getHexagonsByOffsetRange(final int gridXFrom, final int gridXTo, final int gridYFrom, final int gridYTo) {
        final List<CubeCoordinate> coords = new ArrayList<>();

        for (int gridX = gridXFrom; gridX <= gridXTo; gridX++) {
            for (int gridY = gridYFrom; gridY <= gridYTo; gridY++) {
                final int cubeX = CoordinateConverter.convertOffsetCoordinatesToCubeX(gridX, gridY, gridData.getOrientation());
                final int cubeZ = CoordinateConverter.convertOffsetCoordinatesToCubeZ(gridX, gridY, gridData.getOrientation());
                CubeCoordinate coord = CubeCoordinate.fromCoordinates(cubeX, cubeZ);
                if (getByCubeCoordinate(coord).isPresent()) {
                    coords.add(coord);
                }
            }
        }

        final Iterator<CubeCoordinate> coordIter = coords.iterator();

        return new Iterable<Hexagon<T>>() {
            @Override
            public Iterator<Hexagon<T>> iterator() {
                return new Iterator<Hexagon<T>>() {
                    @Override
                    public boolean hasNext() {
                        return coordIter.hasNext();
                    }

                    @Override
                    public void remove() {
                        throw new UnsupportedOperationException("This is not a mutable iterator.");
                    }

                    @Override
                    public Hexagon<T> next() {
                        return getByCubeCoordinate(coordIter.next()).get();
                    }
                };
            }
        };
    }

    @Override
    public boolean containsCubeCoordinate(final CubeCoordinate cubeCoordinate) {
        return this.hexagonDataStorage.containsCoordinate(cubeCoordinate);
    }

    @Override
    public Optional<Hexagon<T>> getByCubeCoordinate(final CubeCoordinate coordinate) {
        return containsCubeCoordinate(coordinate)
                ? Optional.<Hexagon<T>>of(new HexagonImpl<>(gridData, coordinate, hexagonDataStorage))
                : Optional.<Hexagon<T>>empty();
    }

    @Override
    public Optional<Hexagon<T>> getByPixelCoordinate(final double coordinateX, final double coordinateY) {
        int estimatedGridX = (int) (coordinateX / gridData.getHexagonWidth());
        int estimatedGridZ = (int) (coordinateY / gridData.getHexagonHeight());
        estimatedGridX = CoordinateConverter.convertOffsetCoordinatesToCubeX(estimatedGridX, estimatedGridZ, gridData.getOrientation());
        estimatedGridZ = CoordinateConverter.convertOffsetCoordinatesToCubeZ(estimatedGridX, estimatedGridZ, gridData.getOrientation());
        // it is possible that the estimated coordinates are off the grid so we
        // create a virtual hexagon
        final CubeCoordinate estimatedCoordinate = CubeCoordinate.fromCoordinates(estimatedGridX, estimatedGridZ);
        final Hexagon tempHex = new HexagonImpl<>(gridData, estimatedCoordinate, hexagonDataStorage);
        final Hexagon trueHex = refineHexagonByPixel(tempHex, Point.fromPosition(coordinateX, coordinateY));
        if (hexagonsAreAtTheSamePosition(tempHex, trueHex)) {
            return getByCubeCoordinate(estimatedCoordinate);
        } else {
            return containsCubeCoordinate(trueHex.getCubeCoordinate()) ? Optional.<Hexagon<T>>of(trueHex) : Optional.<Hexagon<T>>empty();
        }
    }

    @Override
    public Optional<Hexagon<T>> getNeighborByIndex(final Hexagon<T> hexagon, final int index) {
        final int neighborGridX = hexagon.getGridX() + NEIGHBORS[index][NEIGHBOR_X_INDEX];
        final int neighborGridZ = hexagon.getGridZ() + NEIGHBORS[index][NEIGHBOR_Z_INDEX];
        final CubeCoordinate neighborCoordinate = CubeCoordinate.fromCoordinates(neighborGridX, neighborGridZ);
        return getByCubeCoordinate(neighborCoordinate);
    }

    @Override
    public Collection<Hexagon<T>> getNeighborsOf(final Hexagon<T> hexagon) {
        final Set<Hexagon<T>> neighbors = new HashSet<>();
        for (int i = 0; i < NEIGHBORS.length; i++) {
            final Optional<Hexagon<T>> retHex = getNeighborByIndex(hexagon, i);
            if (retHex.isPresent()) {
                neighbors.add(retHex.get());
            }
        }
        return neighbors;
    }

    private boolean hexagonsAreAtTheSamePosition(final Hexagon<T> hex0, final Hexagon<T> hex1) {
        return hex0.getGridX() == hex1.getGridX() && hex0.getGridZ() == hex1.getGridZ();
    }

    private Hexagon<T> refineHexagonByPixel(final Hexagon<T> hexagon, final Point clickedPoint) {
        Hexagon refined = hexagon;
        double smallestDistance = clickedPoint.distanceFrom(Point.fromPosition(refined.getCenterX(), refined.getCenterY()));
        for (final Hexagon<T> neighbor : getNeighborsOf(hexagon)) {
            final double currentDistance = clickedPoint.distanceFrom(Point.fromPosition(neighbor.getCenterX(), neighbor.getCenterY()));
            if (currentDistance < smallestDistance) {
                refined = neighbor;
                smallestDistance = currentDistance;
            }
        }
        return refined;
    }
}

File: mixite.core/core/src/main/kotlin/org/hexworks/mixite/core/internal/impl/HexagonalGridImpl.kt
package org.hexworks.mixite.core.internal.impl

import org.hexworks.cobalt.datatypes.Maybe
import org.hexworks.mixite.core.api.*
import org.hexworks.mixite.core.api.contract.HexagonDataStorage
import org.hexworks.mixite.core.api.contract.SatelliteData
import org.hexworks.mixite.core.internal.GridData
import kotlin.math.abs

class HexagonalGridImpl<T : SatelliteData>(builder: HexagonalGridBuilder<T>) : HexagonalGrid<T> {

    override val gridData: GridData = builder.gridData
    private val hexagonDataStorage: HexagonDataStorage<T> = builder.getHexagonDataStorage()

    override val hexagons: Iterable<Hexagon<T>>
        get() {
            val coordIter = hexagonDataStorage.coordinates.iterator()

            return object : Iterable<Hexagon<T>> {
                override fun iterator(): Iterator<Hexagon<T>> {
                    return object : Iterator<Hexagon<T>> {
                        override fun hasNext(): Boolean {
                            return coordIter.hasNext()
                        }

                        override fun next(): Hexagon<T> {
                            return hexagon(coordIter.next())
                        }
                    }
                }
            }
        }

    init {
        for (cubeCoordinate in builder.gridLayoutStrategy.fetchGridCoordinates(builder)) {
            this@HexagonalGridImpl.hexagonDataStorage.addCoordinate(cubeCoordinate)
        }
    }

    private fun hexagon(coordinate: CubeCoordinate): HexagonImpl<T> {
        return HexagonImpl(gridData, coordinate, hexagonDataStorage)
    }

    override fun getHexagonsByCubeRange(from: CubeCoordinate, to: CubeCoordinate): Iterable<Hexagon<T>> {
        val coordinates = ArrayList<CubeCoordinate>(abs(from.gridZ-to.gridZ) + abs(from.gridX-to.gridX))

        for (gridZ in from.gridZ..to.gridZ) {
            for (gridX in from.gridX..to.gridX) {
                val coord = CubeCoordinate.fromCoordinates(gridX, gridZ)
                if (containsCubeCoordinate(coord)) {
                    coordinates.add(coord)
                }
            }
        }

        val coordIter = coordinates.iterator()

        return object : Iterable<Hexagon<T>> {
            override fun iterator(): Iterator<Hexagon<T>> {
                return object : Iterator<Hexagon<T>> {
                    override fun hasNext(): Boolean {
                        return coordIter.hasNext()
                    }

                    override fun next(): Hexagon<T> {
                        return hexagon(coordIter.next())
                    }
                }
            }
        }
    }

    override fun getHexagonsByOffsetRange(gridXFrom: Int, gridXTo: Int, gridYFrom: Int, gridYTo: Int): Iterable<Hexagon<T>> {
        val coords = ArrayList<CubeCoordinate>()

        for (gridX in gridXFrom..gridXTo) {
            for (gridY in gridYFrom..gridYTo) {
                val cubeX = CoordinateConverter.convertOffsetCoordinatesToCubeX(gridX, gridY, gridData.orientation)
                val cubeZ = CoordinateConverter.convertOffsetCoordinatesToCubeZ(gridX, gridY, gridData.orientation)
                val coord = CubeCoordinate.fromCoordinates(cubeX, cubeZ)
                if (containsCubeCoordinate(coord)) {
                    coords.add(coord)
                }
            }
        }

        val coordIter = coords.iterator()

        return object : Iterable<Hexagon<T>> {
            override fun iterator(): Iterator<Hexagon<T>> {
                return object : Iterator<Hexagon<T>> {
                    override fun hasNext(): Boolean {
                        return coordIter.hasNext()
                    }

                    override fun next(): Hexagon<T> {
                        return hexagon(coordIter.next())
                    }
                }
            }
        }
    }

    override fun containsCubeCoordinate(coordinate: CubeCoordinate): Boolean {
        return this.hexagonDataStorage.containsCoordinate(coordinate)
    }

    private fun _getByCubeCoordinate(coordinate: CubeCoordinate) =
            if (containsCubeCoordinate(coordinate)) hexagon(coordinate) else null

    override fun getByCubeCoordinate(coordinate: CubeCoordinate): Maybe<Hexagon<T>> =
            Maybe.ofNullable(_getByCubeCoordinate(coordinate))

    override fun getByPixelCoordinate(coordinateX: Double, coordinateY: Double): Maybe<Hexagon<T>> {
        var estimatedGridX = (coordinateX / gridData.hexagonWidth).toInt()
        var estimatedGridZ = (coordinateY / gridData.hexagonHeight).toInt()
        estimatedGridX = CoordinateConverter.convertOffsetCoordinatesToCubeX(estimatedGridX, estimatedGridZ, gridData.orientation)
        estimatedGridZ = CoordinateConverter.convertOffsetCoordinatesToCubeZ(estimatedGridX, estimatedGridZ, gridData.orientation)
        // it is possible that the estimated coordinates are off-grid so we
        // create a virtual hexagon
        val estimatedCoordinate = CubeCoordinate.fromCoordinates(estimatedGridX, estimatedGridZ)
        val centerHex = hexagon(estimatedCoordinate)
        val nearestHex = nearestHexagonToPoint(centerHex, Point.fromPosition(coordinateX, coordinateY))

        return if (nearestHex === centerHex) {
            getByCubeCoordinate(estimatedCoordinate) // centerHex may have been off-grid so look it up again
        } else {
            Maybe.of(nearestHex) // Any other result must be a (real) neighbour
        }
    }

    private fun _getNeighborByIndex(hexagon: Hexagon<T>, index: Int) =
            CubeCoordinate.fromCoordinates(
                    hexagon.gridX + NEIGHBORS[index][NEIGHBOR_X_INDEX],
                    hexagon.gridZ + NEIGHBORS[index][NEIGHBOR_Z_INDEX]
            )

    override fun getNeighborByIndex(hexagon: Hexagon<T>, index: Int) =
            getByCubeCoordinate(_getNeighborByIndex(hexagon, index))

    override fun getNeighborsOf(hexagon: Hexagon<T>): Collection<Hexagon<T>> {
        val neighbors = HashSet<Hexagon<T>>()
        for (i in NEIGHBORS.indices) {
            val retHex = getNeighborByIndex(hexagon, i)
            if (retHex.isPresent) {
                neighbors.add(retHex.get())
            }
        }
        return neighbors
    }

    /*
     * Returns either the original center hex or the nearest (real) hex around it
     */
    private fun nearestHexagonToPoint(centerHex: Hexagon<T>, point: Point): Hexagon<T> {
        var nearest = centerHex
        var nearestDistance = Double.MAX_VALUE
        var current: Hexagon<T>? = nearest // Start with center then check six neighbours

        var i = 0;
        while(true) {
            current?.let {
                val currentDistance = point.distanceFrom(it.center)
                when  {
                    currentDistance < gridData.innerRadius -> return it // Shortcut if well inside bounds of current hex
                    currentDistance < nearestDistance -> {
                        // This covers points right in the corner between three hex's (not within innerRadius of any of them)
                        // TODO In theory, we can shortcut if we have refined twice!
                        nearest = it
                        nearestDistance = currentDistance
                    }
                }
            }

            if (i == 6) {
                return nearest // No direct match, pick the nearest one
            }
            current = _getByCubeCoordinate(_getNeighborByIndex(centerHex, i++))
        }
    }

    companion object {

        private val NEIGHBORS = arrayOf(intArrayOf(+1, 0), intArrayOf(+1, -1), intArrayOf(0, -1), intArrayOf(-1, 0), intArrayOf(-1, +1), intArrayOf(0, +1))
        private const val NEIGHBOR_X_INDEX = 0
        private const val NEIGHBOR_Z_INDEX = 1
    }
}
